rust-runtime/aws-smithy-async/src/future/rendezvous.rs

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+//! Rendezvous channel implementation
+//!
+//! Rendezvous channels are equivalent to a channel with a 0-sized buffer: A sender cannot send
+//! until this is an active receiver waiting. This implementation uses a Semaphore to record demand
+//! and coordinate with the receiver.
+//!
+//! Rendezvous channels should be used with care—it's inherently easy to deadlock unless they're being
+//! used from separate tasks or an a coroutine setup (e.g. [`crate::future::fn_stream::FnStream`])
+
+use std::sync::Arc;
+use std::task::{Context, Poll};
+use tokio::sync::mpsc::error::SendError;
+use tokio::sync::Semaphore;
+
+/// Create a new rendezvous channel
+///
+/// Rendezvous channels are equivalent to a channel with a 0-sized buffer: A sender cannot send
+/// until this is an active receiver waiting. This implementation uses a semaphore to record demand
+/// and coordinate with the receiver.
+pub fn channel<T>() -> (Sender<T>, Receiver<T>) {
+    let (tx, rx) = tokio::sync::mpsc::channel(1);
+    let semaphore = Arc::new(Semaphore::new(0));
+    (
+        Sender {
+            semaphore: semaphore.clone(),
+            chan: tx,
+        },
+        Receiver {
+            semaphore,
+            chan: rx,
+            needs_permit: false,
+        },
+    )
+}
+
+/// Sender-half of a channel
+pub struct Sender<T> {
+    semaphore: Arc<Semaphore>,
+    chan: tokio::sync::mpsc::Sender<T>,
+}
+
+impl<T> Sender<T> {
+    /// Send `item` into the channel.
+    ///
+    /// `send` will block until there is matching demand.
```

Comment:
This doc comment is a bit confusing since the function by definition is non-blocking. Maybe say, "`send` will wake when there is matching demand"?